lib/srv/reexec.go

```diff
@@ -764,39 +665,184 @@ func runForward(handler forwardHandler) (errw io.Writer, code int, err error) {
 	for {
 		buf := make([]byte, 1024)
 		fbuf := make([]*os.File, 1)
-		n, fn, err := uds.ReadWithFDs(conn, buf, fbuf)
+		n, fn, err := uds.ReadWithFDs(parentConn, buf, fbuf)
 		if err != nil {
 			if utils.IsOKNetworkError(err) {
 				return errorWriter, teleport.RemoteCommandSuccess, nil
 			}
 			return errorWriter, teleport.RemoteCommandFailure, trace.Wrap(err)
 		}
-		addr := string(buf[:n])
+
 		if fn == 0 {
-			log.Errorf("Parent did not send a file descriptor for address %q.", addr)
+			slog.ErrorContext(ctx, "Networking request requires a control file.")
 			continue
 		}
 
-		go func() {
-			if err := handler(ctx, addr, fbuf[0]); err != nil {
-				log.WithError(err).Errorf("Error handling forwarding request for address %q.", addr)
-			}
-		}()
+		controlConn, err := uds.FromFile(fbuf[0])
+		_ = fbuf[0].Close()
+		if err != nil {
+			slog.ErrorContext(ctx, "Failed to get a connection from control file.")
+			continue
+		}
+
+		// Starting a new goroutine takes us out of the current PAM context,
+		// so we handle requests synchronously.
+		handleNetworkingRequest(ctx, controlConn, buf[:n])
```

Comment:
This will absolutely destroy the performance of ssh port forwarding.

What needs to happen in the pam thread? The creation of the socket? Or the actual connect/bind/listen?

Reply:
> What needs to happen in the pam thread? The creation of the socket? Or the actual connect/bind/listen?

This is specifically meant to fix pam_namespace, so anything to do with the namespace. Currently that's just creating the temp files (`/tmp/.X11-unix` and /`tmp/teleport-xxx/socket-xxx`) and the `xauth` commands which write to `$HOME/.Xauthority`.

I managed to get a working fix by unlocking the OS thread [which we lock on init for PAM](https://github.com/gravitational/teleport/pull/4133).